File: Valid_BST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        import math
        
        def bst(root,low,high):
            if root is None:
                return True
            
            if root.val>=high or root.val<=low:
                return False
            else:
                return bst(root.left,low,root.val) and bst(root.right,root.val,high)
            
        return bst(root,-math.inf,math.inf)
        

File: ValidBst.java
// Time Complexity : O(n)
// Space Complexity : O(levels)
// Did this code successfully run on Leetcode : YES
// Three line explanation of solution in plain english

class ValidBst {
    TreeNode prev;
    boolean isValid;
    public boolean isValidBST(TreeNode root) {
        prev = null;
        isValid = true;
        Inorder(root);
        return isValid;
    }

    private void Inorder(TreeNode root) {
        if(root == null) return;
        Inorder(root.left);
        if(prev != null && prev.val >= root.val) {
            isValid = false;
        }

        prev = root;
        if(isValid)
            Inorder(root.right);
    }
}